src/derivation/collector/keys_collector.rs

```
@@ -0,0 +1,141 @@
+        preprocessor: KeysCollectorPreprocessor,
+    ) -> Self {
+        let all_factor_sources_in_profile = all_factor_sources_in_profile.into();
+        let (keyrings, factors) = preprocessor.preprocess(all_factor_sources_in_profile);
+
+        let dependencies = KeysCollectorDependencies::new(interactors, factors);
+        let state = KeysCollectorState::new(keyrings);
+
+        Self {
+            dependencies,
+            state: RefCell::new(state),
+        }
+    }
+
+    pub fn new(
+        all_factor_sources_in_profile: IndexSet<HDFactorSource>,
+        derivation_paths: IndexMap<FactorSourceIDFromHash, IndexSet<DerivationPath>>,
+        interactors: Arc<dyn KeysCollectingInteractors>,
+    ) -> Self {
+        let preprocessor = KeysCollectorPreprocessor::new(derivation_paths);
+        Self::with_preprocessor(all_factor_sources_in_profile, interactors, preprocessor)
+    }
+}
+
+impl KeysCollector {
+    fn get_interactor(&self, kind: FactorSourceKind) -> KeyDerivationInteractor {
+        self.dependencies.interactors.interactor_for(kind)
+    }
+
+    /// In decreasing "friction order"
+    async fn derive_with_factors(&self) -> Result<()> {
+        for factors_of_kind in self.dependencies.factors_of_kind.iter() {
+            let interactor = self.get_interactor(factors_of_kind.kind);
+            let client = KeysCollectingClient::new(interactor);
+            client
+                .use_factor_sources(factors_of_kind.factor_sources(), self)
+                .await?;
+        }
+        Ok(())
+    }
+}
+
+impl KeysCollector {
+    fn input_for_interactor(
+        &self,
+        factor_source_id: &FactorSourceIDFromHash,
+    ) -> SerialBatchKeyDerivationRequest {
+        let keyring = self
+            .state
+            .borrow()
+            .keyrings
+            .borrow()
+            .keyring_for(factor_source_id)
+            .unwrap();
+        assert_eq!(keyring.factors().len(), 0);
```

This should throw an error ideally.

---

more of a `precondition` actually, as in, should never happen, and if it does it is programmer error.